x-pack/libbeat/management/managerV2.go

```diff
@@ -311,95 +359,162 @@ func (cm *BeatV2Manager) stopBeat() {
+		case <-cm.reloadCh:
+		default:
+			return
+		}
 	}
+}
 
+func (cm *BeatV2Manager) triggerReload() {
+	cm.drainReload()
+	cm.reloadCh <- struct{}{}
 }
 
-func (cm *BeatV2Manager) handleUnitReload(unit *client.Unit) {
-	cm.addUnit(unit)
-	unitType := unit.Type()
+func (cm *BeatV2Manager) reload(units map[unitKey]*client.Unit) {
+	lowestLevel := client.UnitLogLevelError
+	var outputUnit *client.Unit
+	var inputUnits []*client.Unit
+	var stoppingUnits []*client.Unit
+	for _, unit := range units {
+		state, ll, _ := unit.Expected()
+		if state == client.UnitStateStopped {
+			// unit is being stopped
+			//
+			// we keep the unit so after reload is performed
+			// these units can be marked as stopped
+			stoppingUnits = append(stoppingUnits, unit)
+			continue
+		} else if state != client.UnitStateHealthy {
+			// only stopped or healthy are known (and expected) state
+			// for a unit
+			cm.logger.Errorf("unit %s has an unknown state %+v", unit.ID(), state)
+		}
+		if ll > lowestLevel {
+			lowestLevel = ll
+		}
+		if unit.Type() == client.UnitTypeOutput {
+			outputUnit = unit
+		} else if unit.Type() == client.UnitTypeInput {
+			inputUnits = append(inputUnits, unit)
+		} else {
+			cm.logger.Errorf("unit %s as an unknown type %+v", unit.ID(), unit.Type())
+		}
+	}
 
-	if unitType == client.UnitTypeOutput {
-		cm.handleOutputReload(unit)
-	} else if unitType == client.UnitTypeInput {
-		cm.handleInputReload(unit)
+	// set the new log level (if nothing has changed is a noop)
+	logp.SetLevel(getZapcoreLevel(lowestLevel))
+
+	// reload the output configuration
+	var errs multierror.Errors
+	if err := cm.reloadOutput(outputUnit); err != nil {
+		errs = append(errs, err)
 	}
-}
 
-// Handle the updated config for an output unit
-func (cm *BeatV2Manager) handleOutputReload(unit *client.Unit) {
-	_, _, rawConfig := unit.Expected()
-	if rawConfig == nil {
-		cm.logger.Infof("got output update with no config, ignoring")
-		return
+	// compute the input configuration
+	//
+	// in v2 only a single input type will be started per component, so we don't need to
+	// worry about getting multiple re-loaders (we just need the one for the type)
+	if err := cm.reloadInputs(inputUnits); err != nil {
+		errs = append(errs, err)
 	}
-	cm.logger.Debugf("Got Output unit config '%s'", rawConfig.GetId())
 
-	reloadConfig, err := groupByOutputs(rawConfig)
-	if err != nil {
-		errString := fmt.Errorf("Failed to generate config for output: %w", err)
-		_ = unit.UpdateState(client.UnitStateFailed, errString.Error(), nil)
-		return
+	// report the stopping units as stopped
+	for _, unit := range stoppingUnits {
+		_ = unit.UpdateState(client.UnitStateStopped, "Stopped", nil)
 	}
+
+	// any error during reload changes the whole state of the beat to failed
+	if len(errs) > 0 {
+		cm.status = lbmanagement.Failed
+		cm.message = fmt.Sprintf("%s", errs)
+	}
+
+	// now update the statuses of all units
+	cm.mx.Lock()
+	status := getUnitState(cm.status)
+	message := cm.message
```

Comment:
What's the reasoning behind using the global message? Seems like we should have a status message specifically mentioning the reload operation?

Reply:
This is basically setting the units to Healthy, or what ever beats has set with the `UpdateStatus` for the current state of the beat. The state is global because beats doesn't have a way of passing each unit to each running input and have it handle its own state.